### DISCUSSION TITLE
git: updating your fork to match the source/parent/upstream

### DISCUSSION THREAD

**Discussion by naomi-lgbt** [EddieHubCommunity/support#464] [Tips and Tricks]
Hello~!

You'll likely run in to situations where your forked version of a project "falls behind" the state of the source version. GitHub will allow you to make a pull request to bring the changes from the source into your fork, which *does work*, but creates a merge commit that will show up every time you create a PR back to the source.

Instead, to keep the commit history clean, you can `rebase` your fork against the source. Here are the steps to do so:

- Step one: Make sure you are on your primary branch with `git checkout <branchname>`. Most commonly, this will be `git checkout main`.
- Step two: Get the current state of the source repository with `git fetch upstream`. IF it throws an error that there is no upstream, you'll need to set it ↓
  - Step two.five: To set the upstream, you'll use `git remote add upstream <url>`, where the URL is the clone URL for the source repository.
- Step three: Reset the state of your local repository to match the source with `git reset --hard upstream/<branchname>`, where `branchname` is the name of the primary branch. So, most likely `git reset --hard upstream/main`.
- Step four: Force your fork repo to match this new state with `git push -f`.
- Step five: Celebrate your success!

https://github.com/nhcarrigan/code-snippets/blob/main/git-commands/git-update-fork.sh
I also have the commands saved in my notes, if you'd like to reference them as shell commands.

**Comment by eddiejaoude** [comment]
Thank you @nhcarrigan for this and also for the tips folder 👍  https://github.com/EddieJaoudeCommunity/support/blob/main/tips/git.md
